src/lib/vasi-sync/src/atomic_tls_map.rs

```diff
@@ -0,0 +1,271 @@
+    /// having outstanding references.
+    #[inline]
+    pub unsafe fn get(&self, key: NonZeroUsize) -> Option<Ref<V>> {
+        // SAFETY: Ensured by caller
+        let idx = self.idx(key)?;
+        let ptr = self.values[idx].get();
+        Some(unsafe { Ref::new(ptr, &self.refcounts[idx]) })
+    }
+
+    /// Insert `(key, value)`.
+    ///
+    /// If `key` is already present in `self`, the previous value would shadow
+    /// the newly inserted value. We don't expose this function in the public
+    /// API since this behavior would be confusing.
+    ///
+    /// Returns a reference to the newly inserted value.
+    ///
+    /// # Safety
+    ///
+    /// No other thread must access `self` with `key` in parallel, including
+    /// having outstanding references.
+    unsafe fn insert(&self, key: NonZeroUsize, value: V) -> Ref<V> {
+        let idx = self
+            .indexes_from(key)
+            .find(|idx| {
+                self.keys[*idx]
+                    .compare_exchange(
+                        None,
+                        Some(key),
+                        // Syncs with `Release` on removal
+                        atomic::Ordering::Acquire,
+                        atomic::Ordering::Relaxed,
+                    )
+                    .is_ok()
+            })
+            .unwrap();
+        self.values[idx].get_mut().with(|table_value| {
+            let table_value = unsafe { &mut *table_value };
+            table_value.write(value)
+        });
+        unsafe { Ref::new(self.values[idx].get(), &self.refcounts[idx]) }
+    }
+
+    /// Retrieve the value associated with `key`, initializing it with `init` if `key`
+    /// is not already present.
+    ///
+    /// # Safety
+    ///
+    /// No other thread must access `self` with `key` in parallel, including
+    /// having outstanding references.
+    #[inline]
+    pub unsafe fn get_or_insert_with(&self, key: NonZeroUsize, init: impl FnOnce() -> V) -> Ref<V> {
+        let val = unsafe { self.get(key) };
+        val.unwrap_or_else(|| {
+            let val = init();
+            // SAFETY: Ensured by caller
+            unsafe { self.insert(key, val) }
+        })
+    }
+
+    /// Removes the value still for `key`, if any. Panics if this thread has
+    /// any outstanding references for `key`.
+    ///
+    /// # Safety
+    ///
+    /// *No thread* must have outstanding references to the value at `key`, or perform
+    /// parallel operations on `self` with `key`.
+    pub unsafe fn remove(&self, key: NonZeroUsize) -> Option<V> {
+        let idx = self.idx(key)?;
+        assert_eq!(self.refcounts[idx].get(), 0);
+        let value = self.values[idx].get_mut().with(|value| {
+            let value = unsafe { &mut *value };
+            unsafe { value.assume_init_read() }
+        });
+        // Syncs with `Acquire` on insertion
+        self.keys[idx].store(None, atomic::Ordering::Release);
+        Some(value)
```

Comment:
Maybe comment that anything between the `assume_init_read` above and the `store` must not panic, otherwise we'll drop the value twice.

Reply:
done